core/src/main/java/org/geysermc/geyser/translator/protocol/java/level/JavaSectionBlocksUpdateTranslator.java

```diff
@@ -25,19 +25,190 @@
+import org.geysermc.geyser.util.BlockEntityUtils;
 import org.geysermc.mcprotocollib.protocol.data.game.level.block.BlockChangeEntry;
 import org.geysermc.mcprotocollib.protocol.packet.ingame.clientbound.level.ClientboundSectionBlocksUpdatePacket;
 import org.geysermc.geyser.session.GeyserSession;
 import org.geysermc.geyser.translator.protocol.PacketTranslator;
 import org.geysermc.geyser.translator.protocol.Translator;
 
+import java.util.BitSet;
+
+import static org.geysermc.geyser.level.block.BlockStateValues.JAVA_AIR_ID;
+
 @Translator(packet = ClientboundSectionBlocksUpdatePacket.class)
 public class JavaSectionBlocksUpdateTranslator extends PacketTranslator<ClientboundSectionBlocksUpdatePacket> {
+    private static final int NEIGHBORS_NETWORK_FLAG = (1 << UpdateBlockPacket.Flag.NEIGHBORS.ordinal()) | (1 << UpdateBlockPacket.Flag.NETWORK.ordinal());
+    private static final int NETWORK_FLAG = (1 << UpdateBlockPacket.Flag.NETWORK.ordinal());
 
     @Override
     public void translate(GeyserSession session, ClientboundSectionBlocksUpdatePacket packet) {
+        // Send normal block updates if not many changes
+        if (packet.getEntries().length < 32) {
+            for (BlockChangeEntry entry : packet.getEntries()) {
+                session.getWorldCache().updateServerCorrectBlockState(entry.getPosition(), entry.getBlock());
+            }
+            return;
+        }
+
+        UpdateSubChunkBlocksPacket subChunkBlocksPacket = new UpdateSubChunkBlocksPacket();
+        subChunkBlocksPacket.setChunkX(packet.getChunkX());
+        subChunkBlocksPacket.setChunkY(packet.getChunkY());
+        subChunkBlocksPacket.setChunkZ(packet.getChunkZ());
+
+        // If the entire section is updated, this might be a legacy non-full chunk update
+        // which can contain thousands of unchanged blocks
+        if (packet.getEntries().length == 4096) {
+            // hack - bedrock might ignore the block updates if the chunk was still loading.
+            // sending an UpdateBlockPacket seems to force it
+            BlockChangeEntry firstEntry = packet.getEntries()[0];
+            UpdateBlockPacket blockPacket = new UpdateBlockPacket();
+            blockPacket.setBlockPosition(firstEntry.getPosition());
+            blockPacket.setDefinition(session.getBlockMappings().getBedrockBlock(firstEntry.getBlock()));
+            blockPacket.setDataLayer(0);
+            session.sendUpstreamPacket(blockPacket);
+
+            // Filter out unchanged blocks
+            Vector3i offset = Vector3i.from(packet.getChunkX() << 4, packet.getChunkY() << 4, packet.getChunkZ() << 4);
+            BlockPositionIterator blockIter = BlockPositionIterator.fromMinMax(
+                    offset.getX(), offset.getY(), offset.getZ(),
+                    offset.getX() + 15, offset.getY() + 15, offset.getZ() + 15
+            );
+
+            int[] sectionBlocks = session.getGeyser().getWorldManager().getBlocksAt(session, blockIter);
```

Comment:
This will probably break between integrated and standalone world managers. Standalone will have the cache, Spigot will have the live blocks.